qa/common/shm_util.py

```diff
@@ -344,3 +348,49 @@ def register_add_either_shm_regions(inputs, outputs, shm_region_prefix,
                                          input_byte_size)
         outputs[io_num].set_shared_memory(output_shm_name + '_data',
                                           output_byte_size)
+
+
+class ShmLeakDetector:
+    """Detect shared memory leaks when testing Python backend."""
+
+    class ShmLeakProbe:
+
+        def __init__(self, shm_monitors):
+            self._shm_monitors = shm_monitors
+
+        def __enter__(self):
+            # Wait so that any pending shared memory object is freed.
+            time.sleep(5)
+            self._shm_region_free_sizes = []
+            for shm_monitor in self._shm_monitors:
+                self._shm_region_free_sizes.append(shm_monitor.free_memory())
+
+            return self
+
+        def __exit__(self, type, value, traceback):
+            # Wait so that any pending shared memory object is freed.
+            time.sleep(5)
```

Comment:
Are these sleeps needed only for BLS case? If yes, then they should be called out explicitly in those cases.
No need to slow down other tests with 10sec overhead right? 
+ can we somehow determine there are no pending shared memory operations?

Reply:
> Are these sleeps needed only for BLS case? If yes, then they should be called out explicitly in those cases.
No need to slow down other tests with 10sec overhead right?

The reason for the delay is that the pending shared memory frees, will be called when the backend returns from the ModelInstanceExecute and this is after the responses have been sent. I will put the delay only for the BLS tests.